v5.4.1 upgrade to nightly, tiflash upgrade fail

## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

1. install a tidb cluster with v5.4.1 version
2. upgrade to nightly

### 2. What did you expect to see? (Required)
upgrade successfully

### 3. What did you see instead (Required)
upgrade fail with:
[2022/05/16 08:32:52.849 +08:00] [ERROR] [BaseDaemon.cpp:570] ["BaseDaemon:\n       0x1eba6a1\tfaultSignalHandler(int, siginfo_t*, void*) [tiflash+32220833]\n                \tlibs/libdaemon/src/BaseDaemon.cpp:221\n  0x7f82b6cf9630\t<unknown symbol> [libpthread.so.0+63024]\n       0x7a78a8f\tDB::PageReaderImplMixed::traverse(std::__1::function<void (DB::Page const&)> const&, bool, bool) const [tiflash+128420495]\n                \tdbms/src/Storages/Page/PageStorage.cpp:319\n       0x7938f07\tDB::RegionPersister::restore(DB::TiFlashRaftProxyHelper const*, DB::PageStorage::Config) [tiflash+127110919]\n                \tdbms/src/Storages/Transaction/RegionPersister.cpp:351\n       0x78e7080\tDB::KVStore::restore(DB::TiFlashRaftProxyHelper const*) [tiflash+126775424]\n                \tdbms/src/Storages/Transaction/KVStore.cpp:57\n       0x795e449\tDB::TMTContext::restore(DB::TiFlashRaftProxyHelper const*) [tiflash+127263817]\n                \tdbms/src/Storages/Transaction/TMTContext.cpp:64\n       0x1d437de\tDB::Server::main(std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > const&) [tiflash+30685150]\n                \tdbms/src/Server/Server.cpp:1377\n       0x7fc054a\tPoco::Util::Application::run() [tiflash+133956938]\n                \tcontrib/poco/Util/src/Application.cpp:335\n       0x7fcfd0c\tPoco::Util::ServerApplication::run(int, char**) [tiflash+134020364]\n                \tcontrib/poco/Util/src/ServerApplication.cpp:618\n       0x1d471dd\tmainEntryClickHouseServer(int, char**) [tiflash+30699997]\n                \tdbms/src/Server/Server.cpp:1542\n       0x1cf935e\tmain [tiflash+30380894]\n                \tdbms/src/Server/main.cpp:167\n  0x7f82b6728555\t__libc_start_main [libc.so.6+140629]"] [thread_id=5]


### 4. What is your TiFlash version? (Required)
 ["TiFlash build info: TiFlash\nRelease Version: v6.1.0-alpha\nEdition:         Community\nGit Commit Hash: b8e58ebcb741e3a921fc93352de72c5a20a5ba6f\nGit Branch:      heads/refs/tags/v6.1.0-alpha\nUTC Build Time:  2022-05-13 14:53:47\nEnable Features: jemalloc avx avx512 unwind\nProfile:         RELWITHDEBINFO\n"] [thread_id=1]

--- a/dbms/src/Storages/DeltaMerge/StoragePool.cpp
+++ b/dbms/src/Storages/DeltaMerge/StoragePool.cpp
@@ -357,23 +357,23 @@ PageStorageRunMode StoragePool::restore()
         // If V2 already have no any data in disk, Then change run_mode to ONLY_V3
         if (log_storage_v2->getNumberOfPages() == 0 && data_storage_v2->getNumberOfPages() == 0 && meta_storage_v2->getNumberOfPages() == 0)
         {
-            // TODO: when `compaction` happend
-            // 1. Will rewrite meta into V3 by DT.
-            // 2. Need `DEL` meta in V2.
-            // 3. Need drop V2 in disk and check.
-            LOG_FMT_INFO(logger, "Current pagestorage change from {} to {}", static_cast<UInt8>(PageStorageRunMode::MIX_MODE), static_cast<UInt8>(PageStorageRunMode::ONLY_V3));
+            // TODO: Need drop V2 in disk and check.
+            LOG_FMT_INFO(logger, "Current pagestorage change from {} to {}", //
+                         static_cast<UInt8>(PageStorageRunMode::MIX_MODE),
+                         static_cast<UInt8>(PageStorageRunMode::ONLY_V3));
 
             log_storage_v2 = nullptr;
             data_storage_v2 = nullptr;
             meta_storage_v2 = nullptr;
 
-            log_storage_reader = std::make_shared<PageReader>(run_mode, ns_id, /*storage_v2_*/ nullptr, log_storage_v3, nullptr);
-            data_storage_reader = std::make_shared<PageReader>(run_mode, ns_id, /*storage_v2_*/ nullptr, data_storage_v3, nullptr);
-            meta_storage_reader = std::make_shared<PageReader>(run_mode, ns_id, /*storage_v2_*/ nullptr, meta_storage_v3, nullptr);
+            // Must init by PageStorageRunMode::ONLY_V3
+            log_storage_reader = std::make_shared<PageReader>(PageStorageRunMode::ONLY_V3, ns_id, /*storage_v2_*/ nullptr, log_storage_v3, nullptr);
+            data_storage_reader = std::make_shared<PageReader>(PageStorageRunMode::ONLY_V3, ns_id, /*storage_v2_*/ nullptr, data_storage_v3, nullptr);
+            meta_storage_reader = std::make_shared<PageReader>(PageStorageRunMode::ONLY_V3, ns_id, /*storage_v2_*/ nullptr, meta_storage_v3, nullptr);
 
-            log_storage_writer = std::make_shared<PageWriter>(run_mode, /*storage_v2_*/ nullptr, log_storage_v3);
-            data_storage_writer = std::make_shared<PageWriter>(run_mode, /*storage_v2_*/ nullptr, data_storage_v3);
-            meta_storage_writer = std::make_shared<PageWriter>(run_mode, /*storage_v2_*/ nullptr, meta_storage_v3);
+            log_storage_writer = std::make_shared<PageWriter>(PageStorageRunMode::ONLY_V3, /*storage_v2_*/ nullptr, log_storage_v3);
+            data_storage_writer = std::make_shared<PageWriter>(PageStorageRunMode::ONLY_V3, /*storage_v2_*/ nullptr, data_storage_v3);
+            meta_storage_writer = std::make_shared<PageWriter>(PageStorageRunMode::ONLY_V3, /*storage_v2_*/ nullptr, meta_storage_v3);
 
             max_log_page_id = global_storage_pool->getLogMaxId(ns_id);
             max_data_page_id = global_storage_pool->getDataMaxId(ns_id);

--- a/dbms/src/Storages/Transaction/RegionPersister.cpp
+++ b/dbms/src/Storages/Transaction/RegionPersister.cpp
@@ -317,8 +317,10 @@ RegionMap RegionPersister::restore(const TiFlashRaftProxyHelper * proxy_helper, 
 
             // change run_mode to ONLY_V3
             page_storage_v2 = nullptr;
-            page_writer = std::make_shared<PageWriter>(run_mode, /*storage_v2_*/ nullptr, page_storage_v3);
-            page_reader = std::make_shared<PageReader>(run_mode, ns_id, /*storage_v2_*/ nullptr, page_storage_v3, global_context.getReadLimiter());
+
+            // Must use PageStorageRunMode::ONLY_V3 here.
+            page_writer = std::make_shared<PageWriter>(PageStorageRunMode::ONLY_V3, /*storage_v2_*/ nullptr, page_storage_v3);
+            page_reader = std::make_shared<PageReader>(PageStorageRunMode::ONLY_V3, ns_id, /*storage_v2_*/ nullptr, page_storage_v3, global_context.getReadLimiter());
 
             run_mode = PageStorageRunMode::ONLY_V3;
             break;

